presto-native-execution/presto_cpp/main/TaskManager.cpp

```diff
@@ -559,72 +559,102 @@ std::unique_ptr<TaskInfo> TaskManager::createOrUpdateTaskImpl(
   PromiseHolderWeakPtr<std::unique_ptr<protocol::TaskStatus>> statusRequest;
   PromiseHolderWeakPtr<std::unique_ptr<protocol::TaskInfo>> infoRequest;
 
-  // Create or update task can be called concurrently for the same task.
-  // We need to lock here for allow only one to be executed at a time.
-  // This is especially important for adding splits to the task.
-  std::lock_guard<std::mutex> l(prestoTask->mutex);
+  bool startNextQueuedTask = false;
+  std::unique_ptr<TaskInfo> ret;
+  {
+    // Create or update task can be called concurrently for the same task.
+    // We need to lock here for allow only one to be executed at a time.
+    // This is especially important for adding splits to the task.
+    std::lock_guard<std::mutex> l(prestoTask->mutex);
 
-  if (startTask) {
-    startTaskLocked(prestoTask);
+    if (startTask) {
+      // Default behavior is to start tasks immediately.
+      if (!SystemConfig::instance()->workerOverloadedTaskQueuingEnabled()) {
+        startTaskLocked(prestoTask);
+      } else {
+        if (isServerOverloaded_) {
+          LOG(INFO) << "Server is overloaded. Queueing task " << taskId;
+          taskQueue_.wlock()->emplace(prestoTask);
+        } else {
+          if (!taskQueue_.rlock()->empty()) {
+            LOG(INFO)
+                << "Server is not overloaded, but queued tasks detected. Queueing task "
+                << taskId;
+            auto lockedTaskQueue = taskQueue_.wlock();
+            lockedTaskQueue->emplace(prestoTask);
+            startNextQueuedTask = true;
+          } else {
+            startTaskLocked(prestoTask);
+          }
+        }
+      }
 
-    resultRequests = std::move(prestoTask->resultRequests);
-    statusRequest = prestoTask->statusRequest;
-    infoRequest = prestoTask->infoRequest;
-  }
+      resultRequests = std::move(prestoTask->resultRequests);
+      statusRequest = prestoTask->statusRequest;
+      infoRequest = prestoTask->infoRequest;
+    }
 
-  getDataForResultRequests(resultRequests);
+    getDataForResultRequests(resultRequests);
```

Comment:
Shall we skip doing this if the task get queued?

Reply:
Discussed this.
It looks like we need to do this here.